Copy/Paste text to raw source with click button / keyboard

yipenghe opened the pull request:
continuation of #1975 
(implement) fix #1996 fix #1802 

implement a copy paste mechanism, changing the keyboard shortcut for opening terminal from "c" to "c c" so it does not collide with "a c"

yipenghe committed:
copy attempt, minor bug

yipenghe committed:
copy-dialog

yipenghe committed:
temp checkpt

yipenghe committed:
support copy with keyboard shortcuts

yipenghe committed:
remove log; respect new line char

yipenghe committed:
manual paste instructions

yipenghe committed:
pre-commit

yipenghe committed:
Merge branch 'master' into copy-paste

yipenghe committed:
paste temp

yipenghe committed:
uncheck bundles when click on edit source'

yipenghe committed:
Merge branch 'edit-uncheck' into copy-paste

yipenghe committed:
initial paste draft

yipenghe committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into paste

yipenghe committed:
remove shallow expand-raw-items method, add keyboard support for pasting

yipenghe committed:
clean up comments and console.log

yipenghe committed:
Merge branch 'master' into copy-paste

yipenghe committed:
Merge branch 'copy-paste' into paste

yipenghe committed:
this.focusIndex=>this.state.focusIndex

yipenghe committed:
fix copy after edit source, remove paste part

yipenghe committed:
remove paste part

yipenghe committed:
paste

yipenghe committed:
formatting

yipenghe committed:
remove console.log

yipenghe committed:
Merge branch 'copy-paste' into paste

yipenghe committed:
remove console.log

yipenghe committed:
fix paste click and add tooltip for paste

yipenghe committed:
short cut comments

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
show bundle names in copy dialog for confirmation; uncheck bundles after pasting

yipenghe committed:
Merge branch 'master' into copy-paste

yipenghe committed:
Merge branch 'copy-paste' into paste

yipenghe committed:
exclude invalid bundles when copying

yipenghe committed:
disable paste shortcut when don't have editPerrmision

yipenghe committed:
remove unnecessary saveAndUPdate props for texteditor; fix edit permission problem

yipenghe committed:
size of dialog, paste instruction

yipenghe committed:
Merge branch 'master' into paste

epicfaace committed:
Merge branch 'master' into paste

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Update frontend/src/components/worksheets/WorksheetDialogs.js

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Update frontend/src/components/worksheets/WorksheetDialogs.js

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
remove console.log

yipenghe committed:
rename togglePopup=>toggleCmdDialog; add handleCommand function to reduce code duplication

yipenghe committed:
CallBack=>Callback:

yipenghe committed:
GlossaryModal=>keyboardshortcutModal; add a c/v shortcut help

yipenghe committed:
add worksheet_info mapping info in docstring

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Update codalab/rest/interpret.py

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
remove inline comments & rename

yipenghe committed:
Merge branch 'paste' of https://github.com/codalab/codalab-worksheets into paste

yipenghe commented:
@epicfaace so I found that Firefox actually doesn't support reading from clipboard unless it's a browser extension app. I thinking to change to use localStorage/sessionStorage to remember what was copied, this way we don't need a extra dialog anymore, what are your thoughts on this?

percyliang commented:
It seems like it's nice to actually have it end up on the clipboard, because then it's more interoperable with things outside CodaLab.

yipenghe commented:
> It seems like it's nice to actually have it end up on the clipboard, because then it's more interoperable with things outside CodaLab.

Copy works for both chrome and firefox, but reading from clipboard is not allowed by firefox. We can use local/session storage to complete the pasting (adding to the raw source lines), and still copy the contents to the clipboard as well

percyliang commented:
So if I copy a piece of markdown from a tutorial, I can't paste it in?

yipenghe commented:
> So if I copy a piece of markdown from a tutorial, I can't paste it in?

If we wanna support both Chrome and Firefox, then we can't do that (copy a markdown and paste), because Firefox doesn't allow us to read from the clipboard. I'm looking for some hacks on this

yipenghe committed:
test hack: failing

yipenghe committed:
merge

yipenghe committed:
add paste dialog draft

yipenghe committed:
connect paste dialog with paste mechanism

yipenghe committed:
remove console log

yipenghe committed:
revert ui-tester.py changes

yipenghe committed:
remove extra line

yipenghe commented:
TODO: in pasteboard, add keyboard shortcut to confirm paste directly in text box

epicfaace reviewed (commented):
- When I click on "Pasteboard" and paste, it works properly -- but when I try typing in "a v", nothing happens.
- When I focus on a bundle at the top of a table and paste a bundle, the bundle gets added to the bottom of the table instead of right after the focused bundle.
- We should not show the confirmation dialog when the user copies something empty. This is because, otherwise, it can cause confusing behavior if a user, say, focuses on a bundle row (but doesn't check the box) and presses "a c", and then gets an empty prompt saying "The following bundle ids (excluding invalid ones) will be copied to clipboard:"
- The keyboard shortcut for copy is "a c", but the keyboard shortcut for showing the web terminal is also "c". Thus, a lot of the time in which I am trying to type "a c" to copy something, the web terminal just shows up instead.
- This last point isn't a bug, but more of a general suggestion. I'm not quite comfortable with the idea of a Pasteboard. Isn't the whole idea of the visual editor that you _don't_ have to deal with source lines? Yet in the normal usage of copy/paste, you will have situations like this:

![image](https://user-images.githubusercontent.com/1689183/78898295-8faa7f80-7a41-11ea-944c-726adddea083.png)

In effect, we are having a mini-"Edit Source" feature that is being added here to the visual editor. I think source lines have their place, but that's only in the "Edit Source" page -- the visual editor should try to avoid them if at all possible.

I think it would be good if we could find a more seamless way to copy/paste in which the user _**does not see the source lines at all**_ -- perhaps they could view a preview of the bundle table in the dialog (just like it shows up in the worksheet) instead of the raw source lines?

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Merge branch 'paste' of https://github.com/codalab/codalab-worksheets into paste

yipenghe committed:
use local storage to support copying bundles

yipenghe committed:
clean up code; remove dialogs

yipenghe committed:
Update frontend/src/components/worksheets/ExtraWorksheetHTML/ExtraWorksheetHTML.js

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
Update frontend/src/components/worksheets/ExtraWorksheetHTML/ExtraWorksheetHTML.js

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
Merge branch 'master' into paste

yipenghe commented:
> * When I click on "Pasteboard" and paste, it works properly -- but when I try typing in "a v", nothing happens.
> * When I focus on a bundle at the top of a table and paste a bundle, the bundle gets added to the bottom of the table instead of right after the focused bundle.
> * We should not show the confirmation dialog when the user copies something empty. This is because, otherwise, it can cause confusing behavior if a user, say, focuses on a bundle row (but doesn't check the box) and presses "a c", and then gets an empty prompt saying "The following bundle ids (excluding invalid ones) will be copied to clipboard:"
> * The keyboard shortcut for copy is "a c", but the keyboard shortcut for showing the web terminal is also "c". Thus, a lot of the time in which I am trying to type "a c" to copy something, the web terminal just shows up instead.
> * This last point isn't a bug, but more of a general suggestion. I'm not quite comfortable with the idea of a Pasteboard. Isn't the whole idea of the visual editor that you _don't_ have to deal with source lines? Yet in the normal usage of copy/paste, you will have situations like this:
> 
> ![image](https://user-images.githubusercontent.com/1689183/78898295-8faa7f80-7a41-11ea-944c-726adddea083.png)
> 
> In effect, we are having a mini-"Edit Source" feature that is being added here to the visual editor. I think source lines have their place, but that's only in the "Edit Source" page -- the visual editor should try to avoid them if at all possible.
> 
> I think it would be good if we could find a more seamless way to copy/paste in which the user _**does not see the source lines at all**_ -- perhaps they could view a preview of the bundle table in the dialog (just like it shows up in the worksheet) instead of the raw source lines?

Thanks for all the comments. After changing to use localStorage, everything becomes much cleaner and simpler:
1. No more extra dialogs to confirm copy and paste, I just added a toast info after copying. Since we no longer overwrite users' clipboard, it's also more safer
2. No more source lines! All we do now is select the bundles, copy them, move to the spot we want to paste and then paste
3. since 'c' naturally associates with copy, I just changed the open full terminal shortcut to 'c c' (hit c twice in a row), so it won't trigger the terminal that easily any more

yipenghe commented:
@percyliang At the end I decided to go with local storage, which Ashwin also agrees

Here are some thoughts:

- Supporting bundle copy/cut/paste is what we initially want to support and is the most important thing for this feature. 

- While it would be great to be able to paste Text directly as well, it's not possible to do it directly at this point since firefox restricts the access to user clipboard.
- Thinking about text, one can always click "text" and paste it there, so probably not much value supporting text copy with a different mechanism. (With the dialog approach, we are basically opening another text box when pasting text)

- Moreover, we don't interfere with user clipboard, which I think is a plus as they don't need to worry about having the current clipboard overwritten.

- Last, the process is much simpler since we don't need to open a confirmation dialog

yipenghe committed:
shortcut information update for full terminal

yipenghe committed:
Merge branch 'paste' of https://github.com/codalab/codalab-worksheets into paste

yipenghe committed:
fix casing

yipenghe commented:
> * When I click on "Pasteboard" and paste, it works properly -- but when I try typing in "a v", nothing happens.
> * When I focus on a bundle at the top of a table and paste a bundle, the bundle gets added to the bottom of the table instead of right after the focused bundle.
> * We should not show the confirmation dialog when the user copies something empty. This is because, otherwise, it can cause confusing behavior if a user, say, focuses on a bundle row (but doesn't check the box) and presses "a c", and then gets an empty prompt saying "The following bundle ids (excluding invalid ones) will be copied to clipboard:"
> * The keyboard shortcut for copy is "a c", but the keyboard shortcut for showing the web terminal is also "c". Thus, a lot of the time in which I am trying to type "a c" to copy something, the web terminal just shows up instead.
> * This last point isn't a bug, but more of a general suggestion. I'm not quite comfortable with the idea of a Pasteboard. Isn't the whole idea of the visual editor that you _don't_ have to deal with source lines? Yet in the normal usage of copy/paste, you will have situations like this:
> 
> ![image](https://user-images.githubusercontent.com/1689183/78898295-8faa7f80-7a41-11ea-944c-726adddea083.png)
> 
> In effect, we are having a mini-"Edit Source" feature that is being added here to the visual editor. I think source lines have their place, but that's only in the "Edit Source" page -- the visual editor should try to avoid them if at all possible.
> 
> I think it would be good if we could find a more seamless way to copy/paste in which the user _**does not see the source lines at all**_ -- perhaps they could view a preview of the bundle table in the dialog (just like it shows up in the worksheet) instead of the raw source lines?

About the adding to the end problem, I think it's copies it into the right place unless you are focusing on a bundle row generated by a search. There is focus issue that is not specific to copy/paste: when you add a new run, it adds after the current focused bundle row, but the focus then moves to the last row of the  bundle table. I think it's common focus bug and would be better to address in a different PR.

epicfaace committed:
Fix additem to subworksheets_block and table_block by giving them sort keys

epicfaace committed:
Update frontend/src/components/worksheets/Worksheet/Worksheet.js

epicfaace committed:
prettier and format

epicfaace committed:
remove redunant code

epicfaace committed:
actually, don't need this check

epicfaace committed:
Merge branch 'fix-additem' of github.com:codalab/codalab-worksheets into fix-additem

epicfaace committed:
Merge branch 'master' into fix-additem

epicfaace committed:
Merge branch 'master' into fix-additem

yipenghe committed:
merge master

yipenghe committed:
fix async expanded tables' selection

yipenghe committed:
Merge branch 'fix-additem' into paste

yipenghe committed:
temp

yipenghe committed:
use block to raw for source indexing

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
formatting

yipenghe committed:
Merge branch 'paste' of https://github.com/codalab/codalab-worksheets into paste

yipenghe commented:
turns out that block_to_raw is an accurate mapping from the blocks to the source lines, the only thing that is buggy and in accurate is raw_to_block. I added a mechanism to track which block is a search-expanded one so it would be able to use the right indexing.

Can @epicfaace review?

yipenghe committed:
disable paste when nothing in localStorage, clear localStorage after paste

yipenghe committed:
merge master, resolve conflicts

yipenghe committed:
merge legacy conflicts

yipenghe committed:
fix ids error

yipenghe committed:
remove spaces

yipenghe committed:
remove console.log

yipenghe committed:
wait for small worksheet element to load

yipenghe committed:
formatting

epicfaace committed:
Merge branch 'master' of github.com:codalab/codalab-worksheets into fix-additem

epicfaace committed:
allow adding items right after a bundle row

epicfaace committed:
show new text / new run after each bundle row

epicfaace committed:
fix logic for showNewButtonsAfterEachBundleRow

epicfaace committed:
fix focus when adding after a bundle row

epicfaace committed:
remove changes for RecordsBlockSchema

epicfaace commented:
- when you copy, we should deselect the checkbox immediately (because if you paste right afterwards and the checkbox is still checked, it's confusing)
- when you paste, we shouldn't clear the clipboard
- I think this is still kind of buggy -- for example, I tried having `markdown - markdown - bundle table` and pasted while selecting the first markdown block, then it added the bundle to the bundle table instead of creating a new bundle table after the first markdown block.

https://github.com/codalab/codalab-worksheets/pull/2190 is also ready for you to use with copy/paste, if we want to use it.

yipenghe commented:
> * when you copy, we should deselect the checkbox immediately (because if you paste right afterwards and the checkbox is still checked, it's confusing)
> * when you paste, we shouldn't clear the clipboard
> * I think this is still kind of buggy -- for example, I tried having `markdown - markdown - bundle table` and pasted while selecting the first markdown block, then it added the bundle to the bundle table instead of creating a new bundle table after the first markdown block.
> 
> #2190 is also ready for you to use with copy/paste, if we want to use it.

I think the reason is the add-item end point doesn't update the raw_to_block properly, in this case I will try to use the add-item endpoint, but cut and schema gets trickier

yipenghe committed:
Merge branch 'master' into paste

yipenghe committed:
Merge branch 'master' into fix-additem

yipenghe committed:
add fix-item

yipenghe committed:
fix merge

yipenghe committed:
paste through add-items endpoint

yipenghe committed:
Merge branch 'paste' of https://github.com/codalab/codalab-worksheets into paste

yipenghe committed:
merge latest master

epicfaace committed:
Merge branch 'master' into paste

yipenghe committed:
Update frontend/src/components/worksheets/items/PlaceholderItem.js

Co-authored-by: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
Update frontend/src/components/worksheets/ExtraWorksheetHTML/ExtraWorksheetHTML.js

Co-authored-by: Ashwin Ramaswami <aramaswamis@gmail.com>

yipenghe committed:
remove searchExpanded

yipenghe committed:
Merge branch 'master' into paste

epicfaace committed:
Merge branch 'master' into paste

yipenghe committed:
rm dummy import

yipenghe committed:
undo interpret changes
